### PR TITLE
MDEV-8235: Expose check_slave_start_position as SQL function GTID_CHECK_POS()

### PR DESCRIPTION
**Pull request by ayush-jha123**
I’ve been diving deep into the replication source code for the last two weeks, 
investigating MDEV-8235. I noticed this task has been open for a while and realized 
how useful it would be for DBAs to have a simple way to check GTID availability 
directly from the SQL layer.

After studying how check_slave_start_position() works internally, I’ve implemented 
a user-facing version called GTID_CHECK_POS().

What this PR does:
- Extracted the internal GTID binlog-lookup logic into a shared helper function 
  in sql_repl.cc to avoid code duplication.
- Created the Item_func_gtid_check_pos class to wrap the logic for the SQL parser.
- Added input validation: it returns 1 (exists), 0 (purged), or NULL (invalid input) 
  which makes it robust for automation scripts.
- Included an MTR test main.gtid_check_pos covering standard and edge cases.

I’m really excited to contribute this to MariaDB and would appreciate feedback 
on the implementation.

### CODE REVIEW

File: sql/item_func.h [hunk 1]
```diff
@@ -1505,6 +1505,32 @@ class Item_func_connection_id :public Item_long_func
 };
 
 
+class Item_func_gtid_check_pos :public Item_int_func
```
Comment (gkodinov) [inline sql/item_func.h]: if you are aiming at returning a boolean why not derive from Item_bool_func?

File: sql/sql_repl.cc [hunk 2]
```diff
@@ -2122,6 +2122,44 @@ gtid_state_from_binlog_pos(const char *in_name, uint32 pos, String *out_str)
...
+
+  if (!mysql_bin_log.is_open())
+  {
+    my_error(ER_NO_BINARY_LOGGING, MYF(0));
+    return -1;
+  }
+
+  if (state.load(gtid_str->ptr(), gtid_str->length()))
+  {
+    if (thd && thd->is_error())
```
Comment (gkodinov) [inline sql/sql_repl.cc]: Why clear the error? Why not just return it?

File: sql/sql_repl.cc [hunk 3]
```diff
@@ -2122,6 +2122,44 @@ gtid_state_from_binlog_pos(const char *in_name, uint32 pos, String *out_str)
...
+  */
+  bool found_in_index= false;
+  uint32 out_start_seek= 0;
+  rpl_binlog_state until_binlog_state;
+  until_binlog_state.init();
+  slave_connection_state until_gtid_state;
+  errormsg= gtid_find_binlog_pos(&state, buf, &until_gtid_state, &until_binlog_state, &found_in_index, &out_start_seek);
+  until_binlog_state.free();
+  if (errormsg)
+    return 0;
```
Comment (gkodinov) [inline sql/sql_repl.cc]: This can mean many things. E.g. the log file is absent, cannot be read etc. You will need a proper description of what the returns mean in the JIRA.

File: sql/sql_repl.cc [hunk 4]
```diff
@@ -2122,6 +2122,44 @@ gtid_state_from_binlog_pos(const char *in_name, uint32 pos, String *out_str)
...
+    if (thd && thd->is_error())
+      thd->clear_error();
+    return -1;
+  }
+
+  /*
+    gtid_find_binlog_pos returns an error message if the required
+    GTIDs have been purged from the binary logs.
+  */
+  bool found_in_index= false;
```
Comment (gkodinov) [inline sql/sql_repl.cc:2161]: this is where the count (from the index) gets returned. I guess this is good enough to return. But, the actual use also scans the data. Why aren't you doing anything with it? The function can return empty error message but 0 found GTIDs and your function is returning 1 (a match).

File: sql/sql_repl.cc [hunk 5]
```diff
@@ -2122,6 +2122,44 @@ gtid_state_from_binlog_pos(const char *in_name, uint32 pos, String *out_str)
...
+  char buf[FN_REFLEN];
+  const char *errormsg;
+
+  if (!mysql_bin_log.is_open())
+  {
+    my_error(ER_NO_BINARY_LOGGING, MYF(0));
+    return -1;
+  }
+
+  if (state.load(gtid_str->ptr(), gtid_str->length()))
```
Comment (gkodinov) [inline sql/sql_repl.cc:2138]: This (afaiu) can take many gtids. Not just one. How do you account for that case?

File: sql/item_func.cc [hunk 6]
```diff
@@ -824,6 +825,26 @@ String *Item_int_func::val_str(String *str)
...
 
 
+longlong Item_func_gtid_check_pos::val_int()
+{
+  DBUG_ASSERT(fixed());
+  String *gtid_str= args[0]->val_str(&tmp_value);
+  if ((null_value= args[0]->null_value))
+    return 0;
+
+  int result= gtid_check_pos(current_thd, gtid_str);
```
Comment (gkodinov) [inline sql/item_func.cc:836]: you get a compile error for embedded here. ifdef the whole thing with HAVE_REPLICATION

File: mysql-test/main/gtid_check_pos.test
Comment (gkodinov) [inline mysql-test/main/gtid_check_pos.test:1]: you've forgotten to git add the .result file for this.

File: mysql-test/main/gtid_check_pos.test [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+--echo # 1. Set up some basic GTID history
+CREATE TABLE t1 (a INT);
+INSERT INTO t1 VALUES (1);
+INSERT INTO t1 VALUES (2);
+
+--echo # Get the current GTID position manually and verify it exists
+let $pos= `SELECT @@GLOBAL.gtid_binlog_pos`;
+--replace_result $pos GTID_POS
+eval SELECT GTID_CHECK_POS('$pos');
```
Comment (gkodinov) [inline mysql-test/main/gtid_check_pos.test:15]: you're missing a test for an otherwise valid GTID with the right server and domain id that's absent.

**Review by gkodinov — CHANGES_REQUESTED** [review]
Thank you for your contribution! This is a preliminary review.

First, the formalities:

* please keep a single commit, with a commit message in it.
* please make sure all of the buildbot hosts compile and run successfully

Below are some of my thoughts on the functionality and its implementation and test coverage. You can ignore all of the below and take it with the final reviewer. Or address it and have a faster final review with these fixed.

File: sql/sql_repl.cc [hunk 8]
```diff
@@ -2122,6 +2123,9 @@ gtid_state_from_binlog_pos(const char *in_name, uint32 pos, String *out_str)
 }
 
 
+
```
Comment (gkodinov) [inline sql/sql_repl.cc:2125]: no space only changes please.

File: storage/columnstore/columnstore
Comment (gkodinov) [inline storage/columnstore/columnstore:1]: revert these please

File: libmariadb
Comment (gkodinov) [inline libmariadb:1]: revert this please.

File: sql/item_func.cc [hunk 9]
```diff
@@ -824,6 +825,30 @@ String *Item_int_func::val_str(String *str)
...
+  if ((null_value= args[0]->null_value))
+    return 0;
+
+#ifndef HAVE_REPLICATION
+  my_error(ER_NOT_SUPPORTED_YET, MYF(0), "GTID_CHECK_POS");
+  null_value= 1;
+  return 0;
+#else
+  bool is_reachable= false;
+  if (rpl_gtid_pos_check_reachable(gtid_str, &is_reachable))
```
Comment (gkodinov) [inline sql/item_func.cc:845]: This fails to compile on buildbot (I'm assuming for the unit tests) as follows:

```
/home/buildbot/aarch64-debian-11/build/sql/item_func.cc: In member function ‘virtual bool Item_func_gtid_check_pos::val_bool()’:
/home/buildbot/aarch64-debian-11/build/sql/item_func.cc:841:7: error: ‘rpl_gtid_pos_check_reachable’ was not declared in this scope
  841 |   if (rpl_gtid_pos_check_reachable(gtid_str, &is_reachable))
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Please have a look.

**Review by gkodinov — CHANGES_REQUESTED** [review]
It's looking much better. Please make sure to address the comments below. This is my last batch of preliminary review comments. 
Thanks for your efforts.

**Comment by ayush-jha123** [comment]
Hi @gkodinov,

I've just pushed the updated commit addressing your latest feedback:

Compilaton: Added #include "sql_repl.h" to sql/item_func.cc to fix the scope error you observed on Buildbot.
Submodules: Reverted all changes to the libmariadb and storage/columnstore/columnstore submodules.
Whitespace: I've ensured that no space-only changes are included in the diff; the comment lines and empty lines now match the original source exactly.

Thanks for the review, let me know if there's anything else!

File: mysql-test/main/gtid_check_pos.result [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+# Since we want a deterministic test, let's check
+# a completely fake server_id / seq_no that shouldn't exist
+SELECT GTID_CHECK_POS('0-99-9999999');
+GTID_CHECK_POS('0-99-9999999')
+1
+# 4. Check for an otherwise valid GTID (right domain and server)
+# that is absent (sequence number is in the future).
+SELECT GTID_CHECK_POS('ABSENT_GTID');
+GTID_CHECK_POS('ABSENT_GTID')
+1
```
Comment (gkodinov) [inline mysql-test/main/gtid_check_pos.result:49]: Why is this returning true (1) ? What does this mean? 

Your PR says:
 > it returns 1 (exists), 0 (purged), or NULL (invalid input)

How come an absent GTID exists?

File: mysql-test/main/gtid_check_pos.result [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+PURGE BINARY LOGS BEFORE NOW();
+# Depending on when it is executed, '0-1-1' might still be
+# reachable if we haven't actually purged the very first file
+# but purge binary logs before now() should remove everything 
+# except the active one.
+# Since we want a deterministic test, let's check
+# a completely fake server_id / seq_no that shouldn't exist
+SELECT GTID_CHECK_POS('0-99-9999999');
+GTID_CHECK_POS('0-99-9999999')
+1
```
Comment (gkodinov) [inline mysql-test/main/gtid_check_pos.result:29]: Why is this returning 1? this fake ID exists?

File: sql/item_func.cc [hunk 12]
```diff
@@ -837,6 +826,31 @@ String *Item_int_func::val_str(String *str)
...
 
 
+bool Item_func_gtid_check_pos::val_bool()
+{
+  DBUG_ASSERT(fixed());
+  String *gtid_str= args[0]->val_str(&tmp_value);
+  if ((null_value= args[0]->null_value))
+    return 0;
+
+#ifndef HAVE_REPLICATION
```
Comment (gkodinov) [inline sql/item_func.cc]: I'd reverse the condition for readability. But that's optional.

**Review by gkodinov — CHANGES_REQUESTED** [review]
Thank you for working on this. Some of the test results do not make a lot of sense it seems. 

It'd be good to have even more test coverage: more edge cases, more positive and negative outcomes, scenarios producing all the possible return values and/or errors etc. 

And also a better design definition of what the SQL function does is going to help a lot: e.g. possible return values, when are they returned, formal description or parameters, what determines the return value (just the parameter values or some other state, etc.

File: mysql-test/main/gtid_check_pos.result [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+# 4. Rotate to a new binlog and purge the old one
+FLUSH LOGS;
+# Write a new GTID into the new (post-purge) active binlog
+INSERT INTO t1 VALUES (3);
+# The new position (in the current active log) is reachable
+SELECT GTID_CHECK_POS('NEW_GTID_POS');
+GTID_CHECK_POS('NEW_GTID_POS')
+1
+# The old position (now in a purged log) returns 0
```
Comment (gkodinov) [inline mysql-test/main/gtid_check_pos.result:30]: this one fails on buildbot:

```
main.gtid_check_pos                      w7 [ fail ]
        Test ended at 2026-06-03 19:55:16

CURRENT_TEST: main.gtid_check_pos
--- C:/a/server/server/mysql-test/main/gtid_check_pos.result	2026-06-03 19:12:08.051490800 +0000
+++ C:\a\server\server\mysql-test\main\gtid_check_pos.reject	2026-06-03 19:55:13.752902600 +0000
@@ -30,7 +30,7 @@
 # The old position (now in a purged log) returns 0
 SELECT GTID_CHECK_POS('OLD_GTID_POS');
 GTID_CHECK_POS('OLD_GTID_POS')
-0
+1
```

File: sql/item_func.cc [hunk 14]
```diff
@@ -37,6 +37,7 @@
 #include "sql_acl.h"                            // EXECUTE_ACL
 #include "mysqld.h"                             // LOCK_short_uuid_generator
 #include "rpl_mi.h"
+
```
Comment (gkodinov) [inline sql/item_func.cc]: no white-space only changes please.

File: sql/item_func.cc [hunk 15]
```diff
@@ -837,6 +826,44 @@ String *Item_int_func::val_str(String *str)
...
 
 
+/**
+  @brief Expose replication start position validation as SQL function GTID_CHECK_POS().
+
+  This function accepts a string representation of one or more GTID lists
+  and verifies if they exist and are reachable (viable) within the current
+  set of binary logs.
+
+  @param gtid_str A string representation of GTIDs (e.g. '0-1-1,0-1-2').
```
Comment (gkodinov) [inline sql/item_func.cc]: This will make doxygen return an error if one day it was run on these files. This is a method documentation block.  And yet, you're talking about the SQL function itself. I'd move this particular block towards the class definition (?), reformat it a bit so that there's no @return or @param directives and then add a real one here explaining the return value and what the function does in specific.

File: sql/item_func.cc [hunk 16]
```diff
@@ -837,6 +826,44 @@ String *Item_int_func::val_str(String *str)
...
+
+  @return
```
Comment (gkodinov) [inline sql/item_func.cc]: fwiw, use @retval if you want to do good doxygen. Or don't do doxygen.

File: sql/sql_repl.cc [hunk 17]
```diff
@@ -5838,4 +5839,113 @@ int compare_log_name(const char *log_1, const char *log_2) {
   return res;
 }
 
+/**
```
Comment (gkodinov) [inline sql/sql_repl.cc:5842]: that's what good doxygen looks like! Nice! I'd consider doing some @sa too towards the bigger picture/subsystem. But this is optional.

**Review by gkodinov — CHANGES_REQUESTED** [review]
It is looking much better now! Thanks for the updated tests. This makes it look good. Some final remarks below.

**Comment by ayush-jha123** [comment]
Thnks again for the review, I  pushed an update addressing your final remarks:

1. **Buildbot `w7` failure (`gtid_check_pos.test`)**: 
   The test was failing on Windows because the active binary log was occasionally locked by a background thread, preventing `PURGE BINARY LOGS TO` from successfully removing it before the subsequent `GTID_CHECK_POS()` check was made. I have refactored this block to use `--source include/wait_for_purge.inc` (wrapped in `--disable_result_log`), which safely polls until the log is genuinely purged.
2. **Whitespace only changes**: Removed the empty newline after `#include "rpl_mi.h"` in `item_func.cc`.
3. **Doxygen Cleanup**:
   - I moved the general overview documentation for the SQL function into `item_cmpfunc.h` right above the class definition (and removed `@param` and `@return`).
   - I replaced the documentation block for `val_bool()` in `item_func.cc` with a specific method-level explanation using the proper `@retval` tags.
   - I added the `@sa` cross-references in `sql_repl.cc` to tie it into the bigger picture.

File: mysql-test/main/gtid_check_pos.result (hunk 13 above)
Comment (gkodinov) [reply mysql-test/main/gtid_check_pos.result:30]: It now fails as follows on all platforms:

```
main.gtid_check_pos                      w38 [ fail ]
        Test ended at 2026-06-06 18:32:22
CURRENT_TEST: main.gtid_check_pos
--- /home/buildbot/amd64-msan-clang-20/build/mysql-test/main/gtid_check_pos.result	2026-06-06 18:16:04.000000000 +0000
+++ /home/buildbot/amd64-msan-clang-20/build/mysql-test/main/gtid_check_pos.reject	2026-06-06 18:32:22.403061635 +0000
@@ -21,6 +21,7 @@
 1
 # 4. Rotate to a new binlog and purge the old one
 FLUSH LOGS;
+include/wait_for_purge.inc "master-bin.000003"
 # Write a new GTID into the new (post-purge) active binlog
 INSERT INTO t1 VALUES (3);
 # The new position (in the current active log) is reachable
@@ -30,7 +31,7 @@
 # The old position (now in a purged log) returns 0
 SELECT GTID_CHECK_POS('OLD_GTID_POS');
 GTID_CHECK_POS('OLD_GTID_POS')
-0
+1
 # A completely fake domain/server/seqno (never existed) returns 0
 SELECT GTID_CHECK_POS('99-99-9999999');
 GTID_CHECK_POS('99-99-9999999')
```

File: sql/item_func.h [hunk 18]
```diff
@@ -1507,6 +1505,9 @@ class Item_func_connection_id :public Item_long_func
 };
 
 
+
```
Comment (gkodinov) [inline sql/item_func.h:1508]: some more white-space only changes.

**Review by gkodinov — CHANGES_REQUESTED** [review]
There's still 5 commits: please squash these into 1. 
And there's a test failure across all platforms: please fix.

**Comment by gkodinov** [comment]
Converting to draft, since there's been no reply on my comments for 21 days now. Please convert back to open when/if you intend to resume working on this.
